src/lavinmq/queue/stream_queue.cr

```diff
@@ -0,0 +1,228 @@
+require "./durable_queue"
+require "../client/channel/consumer"
+
+module LavinMQ
+  class StreamQueue < Queue
+    @durable = true
+    @exclusive_consumer = false
+    @no_ack = true
+
+    private def init_msg_store(data_dir)
+      replicator = @vhost.@replicator
+      @msg_store = StreamQueueMessageStore.new(data_dir, replicator)
+    end
+
+    class StreamQueueMessageStore < MessageStore
+      @last_offset = 0_i64
+
+      def initialize(@data_dir : String, @replicator : Replication::Server?)
+        super
+      end
+
+      # Populate bytesize, size and segment_msg_count
+      private def load_stats_from_segments : Nil
+        last_bytesize = 0_u32
+        @segments.each do |seg, mfile|
+          count = 0u32
+          loop do
+            pos = mfile.pos
+            raise IO::EOFError.new if pos + BytesMessage::MIN_BYTESIZE >= mfile.size # EOF or a message can't fit, truncate
+            ts = IO::ByteFormat::SystemEndian.decode(Int64, mfile.to_slice + pos)
+            break mfile.resize(pos) if ts.zero? # This means that the rest of the file is zero, so resize it
+
+            last_bytesize = bytesize = BytesMessage.skip(mfile)
+            count += 1
+            unless deleted?(seg, pos)
+              @bytesize += bytesize
+              @size += 1
+            end
+          rescue ex : IO::EOFError
+            if mfile.pos < mfile.size
+              Log.warn { "Truncating #{mfile.path} from #{mfile.size} to #{mfile.pos}" }
+              mfile.truncate(mfile.pos)
+            end
+            break
+          rescue ex : OverflowError | AMQ::Protocol::Error::FrameDecode
+            raise Error.new(mfile, cause: ex)
+          end
+
+          begin # sets @last_offset to last message offset
+            mfile.seek(mfile.pos - last_bytesize, IO::Seek::Set)
+            msg = BytesMessage.from_bytes(mfile.to_slice + mfile.pos)
+            @last_offset = offset_from_headers(msg.properties.headers)
+          rescue IndexError
+          end
+
+          mfile.pos = 4
+          @segment_msg_count[seg] = count
+        end
+      end
+
+      def shift?(consumer) : Envelope?                              # ameba:disable Metrics/CyclomaticComplexity
+        consumer.update_offset(@last_offset) unless consumer.offset # if no offset provided, use offset of last published message
+        consumer_offset = consumer.offset || 0_i64
+        return if @last_offset <= consumer_offset # don't consume unless there's something to consume
+
+        seg = consumer.segment || @segments.first_value
+        pos = consumer.pos || 0_u32
+        requeued = consumer.requeued || Deque(SegmentPosition).new
+        raise ClosedError.new if @closed
+        if sp = requeued.shift?
+          segment = @segments[sp.segment]
+          begin
+            msg = BytesMessage.from_bytes(segment.to_slice + sp.position)
+            notify_empty(true) if @size.zero?
```

Comment:
no one is intreseted if this consumers "virtual queue" is empty or not

Reply:
yes, already changed with what we discussed earlier today (but not committed yet) 👍